tests/posix/common/src/semaphore.c

```diff
@@ -90,41 +84,189 @@ ZTEST(posix_apis, test_semaphore)
+	zassert_equal(sem_timedwait(sem, &abstime), 0);
 
 	/* TESTPOINT: Semaphore is already acquired, check if
 	 * no semaphore is available
 	 */
-	zassert_equal(sem_timedwait(&sema, &abstime), -1);
+	zassert_equal(sem_timedwait(sem, &abstime), -1);
 	zassert_equal(errno, ETIMEDOUT);
 
-	/* TESTPOINT: sem_destroy with invalid kobject */
-	zassert_equal(sem_destroy(dummy_sem), -1, "invalid"
-		      " semaphore is destroyed");
-	zassert_equal(errno, EINVAL);
-
-	zassert_equal(sem_destroy(&sema), 0, "semaphore is not destroyed");
+	zassert_equal(sem_destroy(sem), 0, "semaphore is not destroyed");
 
 	zassert_equal(pthread_attr_destroy(&attr1), 0,
 		      "Unable to destroy pthread object attrib");
 
 	/* TESTPOINT: Initialize sema with 1 */
-	zassert_equal(sem_init(&sema, 0, 1), 0, "sem_init failed");
-	zassert_equal(sem_getvalue(&sema, &val), 0);
+	zassert_equal(sem_init(sem, 0, 1), 0, "sem_init failed");
+	zassert_equal(sem_getvalue(sem, &val), 0);
 	zassert_equal(val, 1);
 
-	zassert_equal(sem_destroy(&sema), -1, "acquired semaphore"
+	zassert_equal(sem_destroy(sem), -1, "acquired semaphore"
 		      " is destroyed");
 	zassert_equal(errno, EBUSY);
 
 	/* TESTPOINT: take semaphore which is initialized with 1 */
-	zassert_equal(sem_trywait(&sema), 0);
+	zassert_equal(sem_trywait(sem), 0);
 
 	initialize_thread_attr(&attr2);
 
-	zassert_equal(pthread_create(&thread2, &attr2, child_func, NULL), 0,
+	zassert_equal(pthread_create(&thread2, &attr2, child_func, sem), 0,
 		      "Thread creation failed");
 
 	/* TESTPOINT: Wait and acquire semaphore till thread2 gives */
-	zassert_equal(sem_wait(&sema), 0, "sem_wait failed");
+	zassert_equal(sem_wait(sem), 0, "sem_wait failed");
+
+	/* Make sure the threads are terminated */
+	zassert_ok(pthread_join(thread1, NULL));
+	zassert_ok(pthread_join(thread2, NULL));
+}
+
+ZTEST(posix_apis, test_semaphore)
+{
+	sem_t sema;
+
+	/* TESTPOINT: Call sem_post with invalid kobject */
+	zassert_equal(sem_post(NULL), -1, "sem_post of"
+		      " invalid semaphore object didn't fail");
+	zassert_equal(errno, EINVAL);
+
+	/* TESTPOINT: sem_destroy with invalid kobject */
+	zassert_equal(sem_destroy(NULL), -1, "invalid"
+		      " semaphore is destroyed");
+	zassert_equal(errno, EINVAL);
+
+	semaphore_test(&sema);
+}
+
+atomic_val_t sem_get_ref_count(sem_t *sem);
+size_t get_sem_list_len(void);
+#define N_LOOPS 999
+
+ZTEST(posix_apis, test_named_semaphore)
+{
+	sem_t *sem1, *sem2, *different_sem1;
+	size_t num_sem = 0;
+
+	/* If `name` is invalid */
+	sem1 = sem_open(NULL, 0, 0, 0);
+	zassert_equal(errno, EINVAL);
+	zassert_equal(POINTER_TO_INT(sem1), -1);
+	zassert_equal(get_sem_list_len(), num_sem);
+
+	/* Attempt to open a named sem that doesn't exist */
+	sem1 = sem_open("sem1", 0, 0, 0);
+	zassert_equal(errno, ENOENT);
+	zassert_equal(POINTER_TO_INT(sem1), -1);
+	zassert_equal(get_sem_list_len(), num_sem);
+
+	/* Name exceeds CONFIG_SEM_NAMELEN_MAX */
+	sem1 = sem_open("0123456789abcdefg", 0, 0, 0);
```

Comment:
It would be best to modify this so that it uses a (local scope) `static char [CONFIG_SEM_NAMELEN_MAX + 2]`, then fill the string with `a` or something (just remember to terminate with `\0`).

Then it's robust against changes to `CONFIG_SEM_NAMELEN_MAX`.

Reply:
fixed